File: Coin.cpp
#include "Coin.hpp"
#include "Aquarium.hpp"

Coin::Coin(Aquarium& _aquarium, Vector2 _position, int _value) : AquariumObject(_aquarium, 1, _position), value(_value) {
    
}

bool Coin::operator==(const Coin& other) const {
    return this == &other;
}
        
int Coin::getValue() const {
    return value;
}

void Coin::move() {
    if (getPosition().y < getAquarium().getSizeY()) {
        if (getPosition().y + getSpeed() < getAquarium().getSizeY())
            setPosition(getPosition() + Vector2::down * getSpeed());
        else {
            setPosition(Vector2(getPosition().x, getAquarium().getSizeY()));
        }
    }
}

void Coin::destruct() {
    getAquarium().remove(*this);
}

File: Java/src/Coin.java
public class Coin extends AquariumObject implements IDestructible {
  /**
   * Coin's value.
   */
  private final int value;

  /**
   * Constructor.
   * @param aquarium aquarium which will be added by coin
   * @param position position of coin
   * @param value value of this coin
   */
  public Coin(Aquarium aquarium, Vector2 position, int value) {
    super(aquarium, 1, position);
    this.value = value;
  }

  /**
     * Reference comparison.
     * @param other other coin to compare
     * @return true if and only if argument represent this object
     */
  public boolean isEqual(final Coin other) {
    return this == other;
  }

  /**
   * Get value of coin.
   * @return coin's value
   */
  public int getValue() {
    return value;
  }

  /**
   * Move this coin to the bottom.
   */
  public void move() {
    if (getPosition().ordinate < getAquarium().getSizeY()) {
      if (getPosition().ordinate + getSpeed() < getAquarium().getSizeY()) {
        setPosition(getPosition().add(Vector2.down.multiply(getSpeed())));
      } else {
        setPosition(new Vector2(getPosition().abscissa, getAquarium().getSizeY()));
      }
    }
  }

  /**
   * Remove this coin from aquarium.
   */
  public void destruct() {
    getAquarium().remove(this);
  }

  /**
   * Increase money by value.
   */
  public void take() {
    getAquarium().setMoney(getAquarium().getMoney() + getValue());
    destruct();
  }

}